File: packages/urdu-js/src/numbers/words-to-number.ts
import { numberToWords } from './number-to-words.js'

const WORD_MAP = new Map<string, bigint>()

;(function buildMap() {
  for (let i = 0; i <= 99; i++) WORD_MAP.set(numberToWords(BigInt(i)), BigInt(i))
  for (let i = 1; i <= 9; i++) WORD_MAP.set(numberToWords(BigInt(i * 100)), BigInt(i * 100))
  WORD_MAP.set('سو', 100n)
  WORD_MAP.set('ہزار', 1_000n)
  WORD_MAP.set('لاکھ', 100_000n)
  WORD_MAP.set('کروڑ', 10_000_000n)
  WORD_MAP.set('ارب', 1_000_000_000n)
  WORD_MAP.set('کھرب', 1_000_000_000_000n)
  WORD_MAP.set('نیل', 1_000_000_000_000_000n)
})()

export function wordsToNumber(text: string): bigint | null {
  let t = text.trim()
  let negative = false
  if (t.startsWith('منفی ')) { negative = true; t = t.slice(5).trim() }
  const direct = WORD_MAP.get(t)
  if (direct !== undefined) return negative ? -direct : direct
  const tokens = t.split(/\s+/)
  let total = 0n, current = 0n
  for (const tok of tokens) {
    const val = WORD_MAP.get(tok)
    if (val === undefined) return null
    if (val >= 1_000n) { total += (current || 1n) * val; current = 0n }
    else if (val === 100n) { current = (current || 1n) * 100n }
    else current += val
  }
  total += current
  return negative ? -total : total
}

File: packages/urdu-dart/lib/src/numbers/words_to_number.dart
import 'number_to_words.dart';

Map<String, int>? _wordMap;

void _buildMap() {
  if (_wordMap != null) return;
  _wordMap = {};
  for (var i = 0; i <= 99; i++) {
    _wordMap![numberToWords(i)] = i;
  }
  for (var i = 1; i <= 9; i++) {
    _wordMap![numberToWords(i * 100)] = i * 100;
  }
  _wordMap!['سو'] = 100;
  _wordMap!['ہزار'] = 1000;
  _wordMap!['لاکھ'] = 100000;
  _wordMap!['کروڑ'] = 10000000;
  _wordMap!['ارب'] = 1000000000;
  _wordMap!['کھرب'] = 1000000000000;
  _wordMap!['نیل'] = 1000000000000000;
}

/// Converts Urdu words representing a number back to an integer.
int? wordsToNumber(String text) {
  _buildMap();
  var t = text.trim();
  var negative = false;
  if (t.startsWith('منفی ')) {
    negative = true;
    t = t.substring(5).trim();
  }

  final direct = _wordMap![t];
  if (direct != null) return negative ? -direct : direct;

  final tokens = t.split(RegExp(r'\s+'));
  int total = 0, current = 0;
  for (final tok in tokens) {
    final val = _wordMap![tok];
    if (val == null) return null;
    if (val >= 1000) {
      total += (current == 0 ? 1 : current) * val;
      current = 0;
    } else if (val == 100) {
      current = (current == 0 ? 1 : current) * 100;
    } else {
      current += val;
    }
  }
  total += current;
  return negative ? -total : total;
}
